src/dbnode/storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -100,46 +183,82 @@ func (s *commitLogSource) ReadData(
 		return result.NewDataBootstrapResult(), nil
 	}
 
-	readCommitLogPredicate := newReadCommitLogPredicate(
-		ns, shardsTimeRanges, s.opts, s.inspection)
+	var (
+		fsOpts         = s.opts.CommitLogOptions().FilesystemOptions()
+		filePathPrefix = fsOpts.FilePathPrefix()
+	)
 
-	// TODO(rartoul): When we implement caching data across namespaces, this will need
-	// to be commitlog.ReadAllSeriesPredicate() if CacheSeriesMetadata() is enabled
-	// because we'll need to read data for all namespaces, not just the one we're currently
-	// bootstrapping.
-	readSeriesPredicate := newReadSeriesPredicate(ns)
-	iterOpts := commitlog.IteratorOpts{
-		CommitLogOptions:      s.opts.CommitLogOptions(),
-		FileFilterPredicate:   readCommitLogPredicate,
-		SeriesFilterPredicate: readSeriesPredicate,
+	// Determine which snapshot files are available.
+	snapshotFilesByShard, err := s.snapshotFilesByShard(
+		ns.ID(), filePathPrefix, shardsTimeRanges)
+	if err != nil {
+		return nil, err
 	}
+
+	var (
+		bOpts     = s.opts.ResultOptions()
+		blOpts    = bOpts.DatabaseBlockOptions()
+		blockSize = ns.Options().RetentionOptions().BlockSize()
+	)
+
+	// Determine the minimum number of commit logs files that we
+	// must read based on the available snapshot files.
+	readCommitLogPred, mostRecentCompleteSnapshotByBlockShard, err := s.newReadCommitLogPredBasedOnAvailableSnapshotFiles(
+		ns, shardsTimeRanges, snapshotFilesByShard)
+	if err != nil {
+		return nil, err
+	}
+
+	// Setup the commit log iterator.
+	var (
+		nsID              = ns.ID()
+		seriesSkipped     int
+		datapointsSkipped int
+		datapointsRead    int
+
+		// TODO(rartoul): When we implement caching data across namespaces, this will need
+		// to be commitlog.ReadAllSeriesPredicate() if CacheSeriesMetadata() is enabled
+		// because we'll need to read data for all namespaces, not just the one we're currently
+		// bootstrapping.
+		readSeriesPredicate = func(id ident.ID, namespace ident.ID) bool {
+			shouldReadSeries := nsID.Equal(namespace)
+			if !shouldReadSeries {
+				seriesSkipped++
+			}
+			return shouldReadSeries
+		}
+
+		iterOpts = commitlog.IteratorOpts{
+			CommitLogOptions:      s.opts.CommitLogOptions(),
+			FileFilterPredicate:   readCommitLogPred,
+			SeriesFilterPredicate: readSeriesPredicate,
+		}
+	)
+
+	defer func() {
+		s.log.Infof("seriesSkipped: %d", seriesSkipped)
+		s.log.Infof("datapointsSkipped: %d", datapointsSkipped)
+		s.log.Infof("datapointsRead: %d", datapointsRead)
```

Comment:
Should we add a `WithFields(xlog.NewField("bootstrapper": "commitlog"))` or something so we know where this is coming from?

Reply:
Its usually pretty obvious cause of the other logs, but sure I can add that

Reply:
I added it on the struct-level logger so all logs in all these functions get it for free

Reply:
Nice, good stuff.